### DISCUSSION TITLE
Multiple source frames for a single program?

### DISCUSSION THREAD

**Discussion by naren-sriram-robo** [tesseract-robotics/tesseract_ros#166] [Q&A]
Hi, I have a welding like scenario in which I need to pass different source frames for each waypoint and the tool frame (which is the target frame) must align with this source frame. I have my own urdfs, source and target frames. 

I understand this is the part where we pass the source and target frames: 

`CompositeInstruction program(
      "cartesian_program", CompositeInstructionOrder::ORDERED, ManipulatorInfo("manipulator", "grinder_frame", "tool2"));`

I wanted to know if it is possible to pass different source frames as a part of this program or create multiple programs (1 for each waypoint) and pass it to the same request. Any help would be greatly appreciated and I can answer any further questions. 

Thank you.

**Comment by Levi-Armstrong** [comment]
Sure, the one passed to the CompositeInstruction is a global one which is used if a move instruction that does not contain its own ManipulatorInfo. In your case you would assign a manipulator info to each move instruction which has a different working frame.

**Reply by Levi-Armstrong** [reply]
Depending on what you need we may need to discuss how the Simple Planner which just does interpolation should handle interpolating between two state with different working frames. The current functionality is the following.

Original Program:
- State0 (working_fame_a)
- State1 (working_fame_b)

Interpolated Program:
- State0 (working_fame_a)
  - State 0.1a (working_frame_b)
  - State 0.1b (working_frame_b)
  - State 0.1c (working_frame_b)
  - State 0.1d (working_frame_b)
  - State 0.1e (working_frame_b)
- State1 (working_frame_b)

**Reply by naren-sriram-robo** [reply]
Thank you. This is exactly what I was looking for. Yeah, the current interpolation works for me. 
Just a last clarification: Do all these working frames need to be defined in the URDF? Or is there a way I can bypass this and input a transformation instead of defining it in the URDF? I noticed that the 'source_frame' parameter inputs only a string.

**Reply by Levi-Armstrong** [reply]
Yes and no. It is currently not possible in the ManipulatorInfo, but it has been discussed adding a working frame offset and have it work the same way that the tcp_offset works but it has not been added. Though Tesseract environment is fully mutable so you can add a link to the environment pro-grammatically. If you do not want to modify the URDF I would just apply an add link command to the environment and the name given to this link you would use as the working_frame in the ManipulatorInfo.

**Reply by naren-sriram-robo** [reply]
Okay. That works! Thank you for the prompt response.
`add_link_command.cpp` is the reference file for adding links to the environment right? Is there an example of this I can find online or in the source code?

**Reply by Levi-Armstrong** [reply]
Do you need to interact with the environment via ros or do you have access to the environment object?

**Reply by naren-sriram-robo** [reply]
I have access to the environment object.

**Reply by Levi-Armstrong** [reply]
Check out this example and how [this](https://github.com/tesseract-robotics/tesseract_planning/blob/master/tesseract_examples/src/glass_upright_example.cpp#L71) function is used.

**Reply by naren-sriram-robo** [reply]
Thank you Levi! That worked. I was able to add links to the environment and treat them as frames. 
One last unrelated question: This is really the last piece of the puzzle in the problem we are trying to solve. 
So, we want to be able to give a Cartesian waypoint pose this way: (x,y,z,r and p) and let the arm have full freedom in yaw pose. The welding like scenario we are trying to solve is 5 dof and the arm has full freedom in yaw. However, since we need to specify all 6 values of the pose to give a cartesian waypoint, we are unable to add any kind of flexibility in yaw and if we do not give anything, it takes the default value of Identity and if it is unable to reach, the algorithm fails.

`
 Eigen::Quaterniond toolFrameOrient = Eigen::AngleAxisd(0.7,  Eigen::Vector3d::UnitX())
                                     * Eigen::AngleAxisd(-3.14159, Eigen::Vector3d::UnitY())
                                     * Eigen::AngleAxisd(0,   Eigen::Vector3d::UnitZ());           
`
`Waypoint wp1 = CartesianWaypoint(Eigen::Isometry3d::Identity() * Eigen::Translation3d(0.0, 0.0, 0.0) *
                                              toolFrameOrient);`
So my question is: Can we have freedom in yaw or any of the values in the pose while giving a waypoint? That is, in the toolFrameOrient, not have ANY value in UnitZ and have the arm freedom to choose any yaw it wants as long as it reaches the rest of the waypoint?

**Reply by Levi-Armstrong** [reply]
Yes, take a look at the puzzle piece example. This accomplished by defining a profile which gets assigned to the move instruction. Each move instruction has profile and path profile key. The profile key only applies to the waypoint associated with move instruction waypoint and the path profile key apply any point interpolated between this instruction and the previous. Depending one which planner or pipeline you will need to define a profile for each. In the case of using descartes this would be defining a waypoint sampler which generates a set of cartesian pose sampling at some resolution around the z-axis. In the case of trajopt you just need to set the 5th coefficient to zero.

This [file](https://github.com/tesseract-robotics/tesseract_ros_workcell/blob/master/twc_motion_planning/src/twc_planning_server_node.cpp) shows examples for each of the planners creating custom profiles for tool z-axis free and then how you added them to the planning server and uses them when creating the program.

**Reply by naren-sriram-robo** [reply]
Thank you so much. This was really the last thing I wanted and we got it to work for our scenario. We had earlier implemented something similar to Descartes but I think TrajOpt does the job for us. Once again, thank you for being so prompt with responses. 

Just a last follow up question if you do not mind - Where can I check the total cost of each trajectory and for each term info if possible? We need this data for analytics purpose. I was able to get the planning tome by manually adding a clock but total cost and cost per term info would be of great help too. 
Thank you!

**Reply by Levi-Armstrong** [reply]
I don't believe it is exposed in a ros interface, but I would take a look this [print](https://github.com/tesseract-robotics/trajopt/blob/b53d75247a280092b3207bbd2e3ad21c77baa9ae/trajopt_sco/src/optimizers.cpp#L304) function which shows how to extract the costs.

**Reply by naren-sriram-robo** [reply]
Okay sure. Thank you. That helped. 
Just a question out of curiosity: If we want to add tolerance to cartesian constraint, how do you figure we approach the problem? My team and I just took the function 'calcRotationalError()' from utils.cpp and changes the error values to be zero if they are within the tolerance, but that does not seem to work. Do you have any approach in mind or is our understanding of how this error works, wrong?

**Reply by Levi-Armstrong** [reply]
Yea, that probably does not work because the jacobian calculator is not provided. You will want to provide a jocobian calculator which use the error function which does not consider tolerances. This way you always have a valid jacobian. The reason it probably is not working well is because the jacobian is all zeros when you are within tolerance which is not correct.
